doc/neps/nep-0020-gufunc-signature-enhancement.rst

```diff
@@ -0,0 +1,241 @@
+   between operands makes sense. For instance, an ``all_equal`` function that
+   compares vectors in arrays could have a signature ``(n),(n)->()``, but this
+   forces both operands to be arrays, while it would be useful also to check
+   that, e.g., all parts of a vector are constant (maybe zero). The proposal
+   is to allow the implementer of a gufunc to indicate that a dimension can be
+   broadcast by post-fixing the dimension name with ``|1``. Hence, the
+   signature for ``all_equal`` would become ``(n|1),(n|1)->()``.  Another
+   application might be in fused ufuncs, such as ``sumproduct`` (which would
+   have the same signature).
+
+Implementation
+--------------
+
+The proposed changes have all been implemented [3]_, [4]_, [5]_. These PRs
+extend the ufunc structure with two new fields, each of size equal to the
+number of distinct dimensions, with ``core_dim_sizes`` holding possibly fixed
+sizes, and ``core_dim_flags`` holding flags indicating whether a dimension can
+be missing or broadcast.  To ensure we can distinguish between this new
+version and previous versions, an unused entry ``reserved1`` is repurposed as
+a version number.
+
+In the implementation, care is taken that to the elementary function flagged
+dimensions are not treated any differently than non-flagged ones: for
+instance, sizes of fixed-size dimensions are still passed on to the elementary
+function (but the loop can now count on that size being equal to the fixed one
+given in the signature).
+
+An implementation detail to be decided upon is whether it might be handy to
+have a summary of all flags. This could possibly be stored in ``core_enabled``
+(which currently is a bool), with non-zero continuing to indicate a gufunc,
+but specific flags indicating whether or not a gufunc uses fixed, flexible, or
+broadcastable dimensions.
+
+With the above, the formal defition of the syntax would become [4]_::
+
+  <Signature>            ::= <Input arguments> "->" <Output arguments>
+  <Input arguments>      ::= <Argument list>
+  <Output arguments>     ::= <Argument list>
+  <Argument list>        ::= nil | <Argument> | <Argument> "," <Argument list>
+  <Argument>             ::= "(" <Core dimension list> ")"
+  <Core dimension list>  ::= nil | <Core dimension> |
+                             <Core dimension> "," <Core dimension list>
+  <Core dimension>       ::= <Dimension name> <Dimension modifier>
+  <Dimension name>       ::= valid Python variable name | valid integer
+  <Dimension modifier>   ::= nil | "|1" | "?"
+
+#. All quotes are for clarity.
+#. Unmodified core dimensions that share the same name must have the same size.
+   Each dimension name typically corresponds to one level of looping in the
+   elementary function's implementation.
+#. White spaces are ignored.
+#. An integer as a dimension name freezes that dimension to the value.
+#. If a name if suffixed with the "|1" modifier, it is allowed to broadcast
```

Comment:
build is failing, should this be `` not "? Try to do `make html` in `doc/neps` (using python 3.6 for f-strings)

Reply:
Thanks, that was indeed the problem. Plus a few others that are now fixed...